components/script/dom/node.rs

```diff
@@ -2110,13 +2250,19 @@ impl NodeMethods for Node {
             NodeTypeId::CharacterData(..) |
             NodeTypeId::Element(..) |
             NodeTypeId::DocumentType |
-            NodeTypeId::DocumentFragment => Some(self.owner_doc()),
+            NodeTypeId::DocumentFragment(_) => Some(self.owner_doc()),
             NodeTypeId::Document(_) => None,
         }
     }
 
     // https://dom.spec.whatwg.org/#dom-node-getrootnode
-    fn GetRootNode(&self) -> DomRoot<Node> {
+    fn GetRootNode(&self, options: &GetRootNodeOptions) -> DomRoot<Node> {
+        if options.composed {
+            if let Some(shadow_root) = self.rare_data.owner_shadow_root.get() {
+                // shadow-including root.
+                return shadow_root.Host().upcast::<Node>().GetRootNode(options);
+            }
+        }
         self.inclusive_ancestors().last().unwrap()
```

Comment:
If we're in the document this should return the document directly as an optimization.